test/framework/src/main/java/org/elasticsearch/transport/nio/InboundChannelBuffer.java

```diff
@@ -0,0 +1,177 @@
+
+import java.nio.ByteBuffer;
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.function.Supplier;
+
+public class InboundChannelBuffer {
+
+    public static final int PAGE_SIZE = 1 << 14;
+    private final int pageMask;
+    private final int pageShift;
+
+    private final ArrayDeque<Page> pages;
+    private final Supplier<Page> pageSupplier;
+
+    private int capacity = 0;
+    private int internalIndex = 0;
+    private int offset = 0;
+
+    public InboundChannelBuffer() {
+        this(() -> new Page(ByteBuffer.wrap(new byte[PAGE_SIZE]), () -> {}));
+    }
+
+    public InboundChannelBuffer(Supplier<Page> pageSupplier) {
+        this.pageSupplier = pageSupplier;
+        this.pages = new ArrayDeque<>();
+        this.pageMask = PAGE_SIZE - 1;
+        this.pageShift = Integer.numberOfTrailingZeros(PAGE_SIZE);
+        this.capacity = PAGE_SIZE * pages.size();
+        expandCapacity(PAGE_SIZE);
+    }
+
+    public void expandCapacity(int newCapacity) {
+        if (capacity < newCapacity) {
+            int numPages = numPages(newCapacity + offset);
+            int pagesToAdd = numPages - pages.size();
+            for (int i = 0; i < pagesToAdd; ++i) {
+                pages.addLast(pageSupplier.get());
+            }
+            capacity += pagesToAdd * PAGE_SIZE;
+        }
+    }
+
+    public void releasePagesFromHead(int bytesToRelease) {
+        if (bytesToRelease > capacity) {
+            throw new IllegalArgumentException("Releasing more bytes [" + bytesToRelease + "] than buffer capacity [" + capacity + "].");
+        }
+
+        int pagesToRelease = pageIndex(offset + bytesToRelease);
+        for (int i = 0; i < pagesToRelease; ++i) {
+            Page page = pages.removeFirst();
+            page.close();
+        }
+        capacity -= bytesToRelease;
+        internalIndex = Math.max(internalIndex - bytesToRelease, 0);
+        offset = indexInPage(bytesToRelease + offset);
+    }
+
+    public ByteBuffer[] getPreIndexBuffers() {
+        if (internalIndex == 0) {
+            return new ByteBuffer[0];
+        }
+        int indexWithOffset = internalIndex + offset;
+        int pageCount = pageIndex(indexWithOffset);
+        int finalLimit = indexInPage(indexWithOffset);
+        if (finalLimit != 0) {
+            pageCount += 1;
+        }
+
+        ByteBuffer[] buffers = new ByteBuffer[pageCount];
+        Iterator<Page> pageIterator = pages.iterator();
+        ByteBuffer firstBuffer = pageIterator.next().buffer.duplicate();
+        firstBuffer.position(firstBuffer.position() + offset);
+        buffers[0] = firstBuffer;
+        for (int i = 1; i < buffers.length; ++i) {
+            buffers[i] = pageIterator.next().buffer.duplicate();
+        }
+        if (finalLimit != 0) {
+            buffers[buffers.length - 1].limit(finalLimit);
+        }
+
+        return buffers;
+    }
+
+    public ByteBuffer[] getPostIndexBuffers() {
```

Comment:
same here, I need docs.

Reply:
I wonder if it would be easier to have a single method that allows you to get a slice like `ByteBuffer[] slice(long start, long length);` then the caller can decide what is needed? this would remove the need for two methods...